File: cakk-api/src/main/java/com/cakk/api/listener/ViewsIncreaseEventListener.java
package com.cakk.api.listener;

import java.util.Objects;

import org.springframework.scheduling.annotation.Async;
import org.springframework.transaction.event.TransactionPhase;
import org.springframework.transaction.event.TransactionalEventListener;

import lombok.RequiredArgsConstructor;

import com.cakk.api.annotation.ApplicationEventListener;
import com.cakk.domain.mysql.event.views.CakeIncreaseViewsEvent;
import com.cakk.domain.mysql.event.views.CakeShopIncreaseViewsEvent;
import com.cakk.domain.redis.repository.CakeShopViewsRedisRepository;
import com.cakk.domain.redis.repository.CakeViewsRedisRepository;

@RequiredArgsConstructor
@ApplicationEventListener
public class ViewsIncreaseEventListener {

	private final CakeViewsRedisRepository cakeViewsRedisRepository;
	private final CakeShopViewsRedisRepository cakeShopViewsRedisRepository;

	@Async
	@TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
	public void increaseCakeViews(final CakeIncreaseViewsEvent event) {
		cakeViewsRedisRepository.saveOrIncreaseSearchCount(Objects.requireNonNull(event.cakeId()));
	}

	@Async
	@TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
	public void increaseCakeShopViews(final CakeShopIncreaseViewsEvent event) {
		cakeShopViewsRedisRepository.saveOrIncreaseSearchCount(Objects.requireNonNull(event.cakeShopId()));
	}
}

File: cakk-core/src/main/kotlin/com/cakk/core/listener/ViewsIncreaseEventListener.kt
package com.cakk.core.listener


import org.springframework.scheduling.annotation.Async
import org.springframework.transaction.event.TransactionPhase
import org.springframework.transaction.event.TransactionalEventListener

import com.cakk.core.annotation.ApplicationEventListener
import com.cakk.core.dto.event.CakeIncreaseViewsEvent
import com.cakk.core.dto.event.CakeShopIncreaseViewsEvent
import com.cakk.infrastructure.cache.repository.CakeShopViewsRedisRepository
import com.cakk.infrastructure.cache.repository.CakeViewsRedisRepository

@ApplicationEventListener
class ViewsIncreaseEventListener(
    private val cakeViewsRedisRepository: CakeViewsRedisRepository,
    private val cakeShopViewsRedisRepository: CakeShopViewsRedisRepository
) {

	@Async
	@TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
	fun increaseCakeViews(event: CakeIncreaseViewsEvent) {
		cakeViewsRedisRepository.saveOrIncreaseSearchCount(event.cakeId)
	}

	@Async
	@TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
	fun increaseCakeShopViews(event: CakeShopIncreaseViewsEvent) {
		cakeShopViewsRedisRepository.saveOrIncreaseSearchCount(event.cakeShopId)
	}
}
